### PR TITLE
Utility to parametrise those scalars which are only written to once

### PR DESCRIPTION
**Pull request by MichaelSt98**
and assigned a constant value.

**Not really sophisticated, however, this will eventually be superseded by some kind of constant propagation anyway**

E.g.

```fortran
  integer, intent(in) :: invar
  integer, intent(out) :: ret
  integer, parameter :: param = 1
  integer :: a, b, c, d
  real :: x, y, z
  logical :: l1, l2, l3

  ret = 0
  x = 3.5
  y = invar * 2.0
  z = 4.5 + 3.5
  a = ret
  b = 10
  c = invar*invar
  l1 = .false.
  l3 = .true.
  if (c .gt. 0) then
    l2 = .true.
  else
    l2 = .false.
  endif
```

to 

```fortran
  INTEGER, INTENT(IN) :: invar
  INTEGER, INTENT(OUT) :: ret
  INTEGER, PARAMETER :: param = 1
  INTEGER :: a, c, d
  INTEGER, PARAMETER :: b = 10
  REAL :: y
  REAL, PARAMETER :: x = 3.5
  REAL, PARAMETER :: z = 4.5 + 3.5
  LOGICAL :: l2
  LOGICAL, PARAMETER :: l1 = .false.
  LOGICAL, PARAMETER :: l3 = .true.

  ret = 0
  y = invar*2.0
  a = ret
  c = invar*invar
  IF (c > 0) THEN
    l2 = .true.
  ELSE
    l2 = .false.
  END IF
```

### CODE REVIEW

File: loki/transformations/parametrise.py [hunk 1]
```diff
@@ -87,10 +87,60 @@
...
+            return all(issubclass(type(_expr), sym._Literal) for _expr in expr.children)
+        return False
+
+    assignments = FindNodes(ir.Assignment).visit(routine.body)
+    # local variables
+    variables = [var for var in routine.variables if var not in routine.arguments]
+    assignments_dic = {}
+    for assignment in assignments:
+        if assignment.lhs in variables:
+            assignments_dic.setdefault(assignment.lhs, []).append(assignment)
```
Comment (reuterbal) [inline loki/transformations/parametrise.py:119]: I'm not sure but you might be running into case-sensitivity issues here. Could be better to build `assignments_dic` as a `CaseInsensitiveDict` with `assignment.lhs.name` as keys

File: loki/transformations/parametrise.py [hunk 2]
```diff
@@ -87,10 +87,60 @@
...
 
+from loki.transformations.utilities import single_variable_declaration
 from loki.transformations.inline import inline_constant_parameters
 
 
-__all__ = ['ParametriseTransformation']
+__all__ = ['ParametriseTransformation', 'parametrise_single_val_scalars']
+
+
+def parametrise_single_val_scalars(routine):
```
Comment (reuterbal) [inline loki/transformations/parametrise.py:97]: The namings not ideal (as discussed offline). In particular, `parametrise` in the transformation refers to replacing values by constant parameters, whereas here we essentially mark local variables as compile-time constants. How about `declare_fixed_value_scalars_as_constants`?

File: loki/transformations/tests/test_parametrise.py [hunk 3]
```diff
@@ -473,3 +473,57 @@ def test_parametrise_non_driver_entry_points(tmp_path, testdir, frontend, config
...
+  z = 4.5 + 3.5
+  a = ret
+  b = 10
+  c = invar*invar
+  l1 = .false.
+  l3 = .true.
+  if (c .gt. 0) then
+    l2 = .true.
+  else
+    l2 = .false.
```
Comment (reuterbal) [inline loki/transformations/tests/test_parametrise.py:502]: Try upper-casing the lhs here to make sure my comment above is redundant ;-)

File: loki/transformations/parametrise.py [hunk 4]
```diff
@@ -87,10 +87,60 @@
...
+    """
+    Parametrise those scalars which are only written to once and assigned a constant value.
```
Comment (reuterbal) [inline loki/transformations/parametrise.py:99]:
```suggestion
    Mark local scalars that are assigned a fixed value as parameters
```

**Review by reuterbal — CHANGES_REQUESTED** [review]
A few minor comments to make sure this handles a few edge cases better, otherwise I think this is nice and simple!

**Review by reuterbal — APPROVED** [review]
Many thanks for the expanded test case and addressing the comments.

**Review by mlange05 — APPROVED** [review]
Looks great! GTG from me. :shipit:
